agent/fw_drupal8.c

```diff
@@ -436,40 +488,58 @@ NR_PHP_WRAPPER(nr_drupal94_invoke_all_with) {
     goto leave;
   }
 
-  prev_hook = NRPRG(drupal_module_invoke_all_hook);
-  prev_hook_len = NRPRG(drupal_module_invoke_all_hook_len);
-  NRPRG(drupal_module_invoke_all_hook)
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO \
+    && !defined OVERWRITE_ZEND_EXECUTE_DATA
+  invoke_all_push_hook_stacks(hook);
+#else
+  prev_hook = NRPRG(drupal_invoke_all_hook);
+  prev_hook_len = NRPRG(drupal_invoke_all_hook_len);
+  NRPRG(drupal_invoke_all_hook)
       = nr_strndup(Z_STRVAL_P(hook), Z_STRLEN_P(hook));
-  NRPRG(drupal_module_invoke_all_hook_len) = Z_STRLEN_P(hook);
-
+  NRPRG(drupal_invoke_all_hook_len) = Z_STRLEN_P(hook);
+#endif // OAPI
   callback = nr_php_arg_get(2, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+
   /* This instrumentation will fail if callback has already been wrapped
    * with a special instrumentation callback in a different context.
    * In this scenario, we will be unable to instrument hooks and modules for
    * this particular call */
-
-  /*
-   * Temporary exclusion via #if defined OVERWRITE_ZEND_EXECUTE_DATA
-   * This needs to be excluded for now because the hooks handling was changed
-   * for oapi so the hook vars this is referencing do not exist.
-   *
-   */
-#if defined OVERWRITE_ZEND_EXECUTE_DATA
   nr_php_wrap_generic_callable(callback,
                                nr_drupal94_invoke_all_with_callback TSRMLS_CC);
-#endif
   NR_PHP_WRAPPER_CALL;
 
   nr_php_arg_release(&callback);
-  nr_free(NRPRG(drupal_module_invoke_all_hook));
-  NRPRG(drupal_module_invoke_all_hook) = prev_hook;
-  NRPRG(drupal_module_invoke_all_hook_len) = prev_hook_len;
-
-leave:
+#if ZEND_MODULE_API_NO < ZEND_8_0_X_API_NO \
+    || defined OVERWRITE_ZEND_EXECUTE_DATA
+  nr_free(NRPRG(drupal_invoke_all_hook));
+  NRPRG(drupal_invoke_all_hook) = prev_hook;
+  NRPRG(drupal_invoke_all_hook_len) = prev_hook_len;
+#endif // not OAPI
+
+leave: ;
+#if ZEND_MODULE_API_NO < ZEND_8_0_X_API_NO \
+    || defined OVERWRITE_ZEND_EXECUTE_DATA
+  /* for OAPI, the _after callback handles this free */
   nr_php_arg_release(&hook);
+#endif // not OAPI
+}
+NR_PHP_WRAPPER_END
+
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO \
+    && !defined OVERWRITE_ZEND_EXECUTE_DATA
+NR_PHP_WRAPPER(nr_drupal94_invoke_all_with_after) {
+  (void)wraprec;
+  invoke_all_clean_hook_stacks();
```

Comment:
Should the 'after' callback cleanup the whole stack or just pop the last pushed item? Pre-oapi seemed to do the later, i.e. pop the last item from the stack.

Reply:
This call does just pop the last elements of the stack, explained more thoroughly here: https://github.com/newrelic/newrelic-php-agent/pull/701#discussion_r1276267265.

"clean" is perhaps the wrong word to use. Maybe "pop" instead?

Reply:
Yes, 'pop' as a verb would make things more obvious. If you consider re-naming the 'stacks' API, please consider [this suggestion](https://github.com/newrelic/newrelic-php-agent/pull/701#discussion_r1275366315) for naming the functions.